### PR TITLE
Create ResolutionParameters

### PR DESCRIPTION
**Pull request by jvandort**
ResolutionParameters are a mostly-immutable but completely thread-safe set of properties used as input to a dependency resolution. The goal is to have these parameters be the complete set of project-scoped state required to perform a dependency resolution. Furthermore, this state should remain safe to interact with without a project lock, meaning once we migrate all project-derived parameters to this class we can perform dependency resolution without acquiring a project lock. 

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationSpec.groovy
Comment (jvandort) [inline platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationSpec.groovy:1668]: Can we add any of these back in some form?

Comment (jvandort) [reply platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationSpec.groovy:1668]: No we cannot

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/ComponentModuleMetadataContainer.java
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/ComponentModuleMetadataContainer.java:1]: Logic moved to `DefaultComponentModuleMetadataHandler.java`

**Comment by jvandort** [comment]
@bot-gradle test prf apt

**Comment by jvandort** [comment]
@bot-gradle test prf apt

**Comment by jvandort** [comment]
@bot-gradle test prf apt

**Comment by jvandort** [comment]
@bot-gradle test prf apt

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ResolutionExecutor.java
Comment (ghale) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ResolutionExecutor.java:565]: ❓ Is this change necessary?  This feels like a step backwards from the previous message (i.e. "dependency was locked to version ${lockVersion}..."). This new message may not be meaningful to users.  At least including the version makes it clear what value is in effect.  Maybe I'm not understanding, but don't we have everything we need here to construct a similar message to what we had before?

**Review by ghale — APPROVED** [review]
Overall, this feels like a good change.  

I left one question (about the dependency lock reason message) that I think should at least be answered before we merge this (even if the answer is we leave it as is).

Otherwise, there are a few other comments/suggestions which shouldn't block merging (i.e. take em or leave em).

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ResolutionExecutor.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ResolutionExecutor.java:565]: I agree this is in a way a step backwards -- but I think it is quite a minor one -- the reason being that the version is always very close by. Consider the output of the `dependencyInsight` task where this reason is visible (on master -- without this change):

```
> Task :dependencyInsight
org:foo:1.0
  Variant default:
    | Attribute Name    | Provided | Requested |
    |-------------------|----------|-----------|
    | org.gradle.status | release  |           |
   Selection reasons:
      - By constraint: dependency was locked to version '1.0'
```

The version is just a few lines above the printed reason.

Same goes for this error message about dependency constraints that cannot be satisfied -- the version is present on the same line:

```
      > Cannot find a version of 'org:foo' that satisfies the version constraints:
           Dependency path ':depLock:unspecified' --> 'org:foo:1.+'
           Dependency path ':depLock:unspecified' --> 'org:foo:{strictly 1.1}'
           Constraint path ':depLock:unspecified' --> 'org:foo:{strictly 1.0}' because of the following reason: dependency was locked to version '1.0'
```

Finally, the version is readily accessible programmatically, right next to the reason, in `ResolvedComponentResult`. You get the reason with `getSelectionReason`, and you get the version from the `getId` method of a `ComponentResult` (the parent interface). 

Now, _why_ did I did I make this change? Its not really related to this PR other than I touched the code and felt I should make the change -- but this causes us to allocate a new String for each component of each dependency graph when dependency locking is enabled. And, these strings are retained in memory for the entire build during sync.

Say we have a 1000 project build, each with ~200 configurations, lets say conservatively 30 resolvable configurations with 10 dependencies each (thats a really small graph!) At 36 bytes per string (at minimum), thats 10MB. Scale up to an average of 100 components per graph and 5000 projects like in some of the larger builds we're seeing and thats 500MB of Strings. 

I recall seeing this at some point when I was analyzing heap dumps.

File: platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ShortCircuitingResolutionExecutorSpec.groovy
Comment (jvandort) [reply platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ShortCircuitingResolutionExecutorSpec.groovy:41]: Makes sense. 

Switched to `def resolveContext = Mock(ResolveContext)`

File: platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/dsl/ComponentModuleMetadataContainerTest.groovy
Comment (jvandort) [reply platforms/software/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/dsl/ComponentModuleMetadataContainerTest.groovy:33]: Renamed to `modules`

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/GlobalDependencyResolutionRules.java
Comment (jvandort) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/GlobalDependencyResolutionRules.java:1]: I think like everything there is a balance between writing an interface for everything and never using interfaces. 

For types exposed publicly I think the scale tips almost entirely to always using interfaces. 

For internal types, even services, I think I would argue writing an interface for everything is superfluous. There are certainly many valid use cases for writing interfaces, but I also think Java's `public/private` separation is powerful enough for types without interfaces to hide the methods they do not want their users to call. 

In any case I think this class is simple enough to not warrant an interface itself. In fact it is a small wrapper around an interface -- `DependencySubstitutionRules` -- but cannot implement that interface itself as it also _injects_ a list of those interfaces. This would otherwise cause an error in the service registry infrastructure as it does not allow you to inject a list of services where you implement one of those services. That probably points to this code needing some rework, though I would argue this is cleaner than before. 

I've certainly introduced services with more complex functionality with this -- without an interface -- so if this is something you think we should avoid for more complex things its probably worth talking about this more so we can 1. avoid digging ourselves into a deeper hole and 2. work towards restricting this sort of pattern in a more programmatic fashion (i.e. not allowing you to declare services that don't have an interface)

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/GlobalDependencyResolutionRules.java
Comment (ghale) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/GlobalDependencyResolutionRules.java:1]: I 100% agree there's a balance that should be struck and that this class is an example of one that is probably fine as just an implementation class.  

I think there are a few cases where using an interface/implementation separation obviously makes sense:

- public API
- services that could be used by other platforms
- services that could be used by other architectural modules in the same platform
- services with multiple interchangeable implementations
- services with multiple implementations that are consumed in aggregate
- services with cyclical relationships

There might be a few I'm not thinking of, too.  But I think outside of those cases, it's probably a "whatever seems right" situation.

I reckon this is something that we should raise in an ADR.  I'll discuss it with the AC.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ResolutionExecutor.java
Comment (ghale) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ResolutionExecutor.java:565]: The memory concerns seem compelling, and I see what you mean about the version being nearby.  It does mean the message is sort of useless outside of the rest of the info, but that's a very minor nitpick.  

I do feel like we could improve the message, but I don't have any ideas that don't seem similarly vague:
- "the dependency was locked to this version"
- "the dependency version is locked"
- "a dependency lock enforces this version"

Just thinking out loud, I wonder if we could address the memory issue with the original message by using a constant format value and only creating the string when it's actually rendered.  i.e. `dependency was locked to version '%s'` and `getReason()` calls `String.format()` somehow.  Just sharing the thought - not suggesting we need to run off and investigate this.

**Pull request merged**
